src/interop/java/Full/Random.java

```diff
@@ -0,0 +1,104 @@
+package DafnyVMC;
+
+import dafny.TypeDescriptor;
+import java.math.BigInteger;
+import java.security.SecureRandom;
+
+public class Random implements DafnyVMCTrait.RandomTrait {
+  static ThreadLocal<SecureRandom> rng;
+  
+  public Random() {
+    this.rng = ThreadLocal.withInitial(Random::createSecureRandom);
+  }
+
+  public Random(SecureRandom rng) {
+    this.rng = ThreadLocal.withInitial(() -> rng);
+  }
+
+  private static final SecureRandom createSecureRandom() {
+    final SecureRandom rng = new SecureRandom();
+    // Required for proper initialization
+    rng.nextBoolean(); 
+    return rng;
+  }
+
+  public BigInteger UniformPowerOfTwoSample(BigInteger n) {
+    if (n.compareTo(BigInteger.ONE) < 0) {
+      throw new IllegalArgumentException("n must be positive");
+    }
+
+    return new BigInteger(n.bitLength()-1, rng.get());
+  }
+
+  public java.math.BigInteger UniformIntervalSample(java.math.BigInteger a, java.math.BigInteger b) {
+    return Uniform.Interface._Companion_Trait.UniformIntervalSample(this, a, b);
+  }
+
+  public java.math.BigInteger UniformSample(java.math.BigInteger n) {
+    return Uniform.Implementation._Companion_Trait.UniformSample(this, n);
+  }
+
+  public boolean CoinSample() {
```

Comment:
A problem with doing this is that the Java API will be nicer, but is the idea that we should do this for every target?